File: CSES/1. Introductory Problems/CSES-1092_Two Sets.py
"""
    Greedy
    先算出 [1, n] 的和 s，如果 s 是奇數，則無解；否則可以分成兩個和為 s//2 的集合。
    由大到小枚舉每個數字，如果加上這個數字後和不超過 s//2，則加入第一個集合，否則加入第二個集合。
    最後輸出兩個集合。
"""
n = int(input())
s = n * (n + 1) // 2 # sum of n numbers
s1 = s2 = s // 2
res1, res2 = [], []

if s & 1:
    print("NO")
else:
    for x in range(n, 0, -1):
        if s1 - x >= 0:
            res1.append(x)
            s1 -= x
        else:
            res2.append(x)
            s2 -= x
    print("YES")
    print(len(res1))
    print(*res1)
    print(len(res2))
    print(*res2)

File: CSES/1. Introductory Problems/CSES-1092_Two Sets.cpp
#include <bits/stdc++.h>
using namespace std;
using LL = long long;
#define endl '\n'

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    LL n; cin >> n;
    LL s = n * (n + 1) / 2;
    LL s1 = s / 2, s2 = s / 2;
    vector<int> res1, res2;
    if (s & 1) {
        cout << "NO" << endl;
    }
    else {
        for (int x = n; x > 0; x--) {
            if (s1 - x >= 0) {
                res1.push_back(x);
                s1 -= x;
            }
            else {
                res2.push_back(x);
                s2 -= x;
            }
        }
        cout << "YES" << endl;
        cout << res1.size() << endl;
        for (int i = 0; i < res1.size(); i++) {
            cout << res1[i] << (i == res1.size() - 1 ? endl : ' ');
        }
        cout << res2.size() << endl;
        for (int i = 0; i < res2.size(); i++) {
            cout << res2[i] << (i == res2.size() - 1 ? endl : ' ');
        }
    }
    return 0;
}